dynawo/sources/Modeler/Common/DYNDynamicData.cpp

```diff
@@ -274,6 +275,36 @@ DynamicData::getParametersSet(const string& modelId, const string& parFile, cons
   return parametersSetCollection->getParametersSet(parId);
 }
 
+void
+DynamicData::resolveParReferences(shared_ptr<Model> model, shared_ptr<ParametersSet> modelSet) {
+  if (modelSet == nullptr)
+    return;
+  boost::unordered_map<std::string, boost::shared_ptr<Reference> >& references = modelSet->getReferences();
+  for (const std::pair<std::string, boost::shared_ptr<Reference> >& ref : references) {
+    const shared_ptr<Reference>& parRef = ref.second;
+    if (parRef->getOrigData() == Reference::OriginData::PAR) {
+      std::string parFile;
+      if (parRef->getParFile().empty()) {
+        parFile = modelSet->getFilePath();
+      } else {
+        parFile = parRef->getParFile();
+      }
+      std::string parFilePath = canonical(parFile, rootDirectory_);
+      shared_ptr<ParametersSet> referencedParametersSet = getParametersSet(model->getId(), parFilePath, parRef->getParId());
+      shared_ptr<Parameter> referencedParameter = referencedParametersSet->getParameter(parRef->getOrigName());
+      if (parRef->getType() == "DOUBLE") {
```

Comment:
This might segfault if the parameter does not exist

Reply:
The type_ attribute is  default initialized by the Reference constructor so there is no undefined behaviour possible.

Reply:
I'm not sure it's possible to test DynamicData::resolveParReferences method since it's a private method.

Reply:
parRef is a pointer to Reference class. It comes from the references_ map which is filled in by Xerces.

> > The type_ attribute is default initialized by the Reference constructor so there is no undefined behaviour possible.
> 
> but parRef can be nullptr no?

Reply:
sorry I meant referencedParameter

Reply:
If the parId is wrong, the ParametersSetNotFound exception is thrown.
If the file containing the referenced value doesn't exist, the parsing will fail.
If the referenced value doesn't exist, the ParameterNotFoundInSet exception is thrown.
If the reference and the referenced value have different types, the ParameterInvalidTypeRequested exception is thrown.